// agentql/dist/ext/playwright/driver-constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_INCLUDE_HIDDEN_ELEMENTS = exports.DEFAULT_INCLUDE_HIDDEN_DATA = exports.DEFAULT_WAIT_FOR_NETWORK_IDLE = exports.DEFAULT_QUERY_DATA_TIMEOUT_SECONDS = exports.DEFAULT_QUERY_ELEMENTS_TIMEOUT_SECONDS = exports.SLOWEST_WEBSITE_AVG_LOAD_TIME_SECONDS = exports.WEBSITE_AVG_LOAD_TIME_SECONDS = void 0;
const WEBSITE_AVG_LOAD_TIME_SECONDS = 6;
exports.WEBSITE_AVG_LOAD_TIME_SECONDS = WEBSITE_AVG_LOAD_TIME_SECONDS;
const SLOWEST_WEBSITE_AVG_LOAD_TIME_SECONDS = 14;
exports.SLOWEST_WEBSITE_AVG_LOAD_TIME_SECONDS = SLOWEST_WEBSITE_AVG_LOAD_TIME_SECONDS;
const DEFAULT_QUERY_ELEMENTS_TIMEOUT_SECONDS = 300;
exports.DEFAULT_QUERY_ELEMENTS_TIMEOUT_SECONDS = DEFAULT_QUERY_ELEMENTS_TIMEOUT_SECONDS;
const DEFAULT_QUERY_DATA_TIMEOUT_SECONDS = 900;
exports.DEFAULT_QUERY_DATA_TIMEOUT_SECONDS = DEFAULT_QUERY_DATA_TIMEOUT_SECONDS;
const DEFAULT_WAIT_FOR_NETWORK_IDLE = true;
exports.DEFAULT_WAIT_FOR_NETWORK_IDLE = DEFAULT_WAIT_FOR_NETWORK_IDLE;
const DEFAULT_INCLUDE_HIDDEN_DATA = true;
exports.DEFAULT_INCLUDE_HIDDEN_DATA = DEFAULT_INCLUDE_HIDDEN_DATA;
const DEFAULT_INCLUDE_HIDDEN_ELEMENTS = false;
exports.DEFAULT_INCLUDE_HIDDEN_ELEMENTS = DEFAULT_INCLUDE_HIDDEN_ELEMENTS;
